Source/inv_iterators.hpp

```diff
@@ -0,0 +1,297 @@
+	{
+	}
+
+	[[nodiscard]] ItemsContainerRange::Iterator begin() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerRange::Iterator { &player_->InvBody[0], ContainerSize(), 0 };
+	}
+
+	[[nodiscard]] ItemsContainerRange::Iterator end() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerRange::Iterator { nullptr, ContainerSize(), ContainerSize() };
+	}
+
+private:
+	[[nodiscard]] std::size_t ContainerSize() const
+	{
+		return sizeof(player_->InvBody) / sizeof(player_->InvBody[0]);
+	}
+
+	PlayerStruct *player_;
+};
+
+/**
+ * @brief A range over non-equipped inventory player items.
+ */
+class InventoryPlayerItemsRange {
+public:
+	explicit InventoryPlayerItemsRange(PlayerStruct &player)
+	    : player_(&player)
+	{
+	}
+
+	[[nodiscard]] ItemsContainerRange::Iterator begin() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerRange::Iterator { &player_->InvList[0], ContainerSize(), 0 };
+	}
+
+	[[nodiscard]] ItemsContainerRange::Iterator end() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerRange::Iterator { nullptr, ContainerSize(), ContainerSize() };
+	}
+
+private:
+	[[nodiscard]] std::size_t ContainerSize() const
+	{
+		return static_cast<std::size_t>(player_->_pNumInv);
+	}
+
+	PlayerStruct *player_;
+};
+
+/**
+ * @brief A range over belt player items.
+ */
+class BeltPlayerItemsRange {
+public:
+	explicit BeltPlayerItemsRange(PlayerStruct &player)
+	    : player_(&player)
+	{
+	}
+
+	[[nodiscard]] ItemsContainerRange::Iterator begin() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerRange::Iterator { &player_->SpdList[0], ContainerSize(), 0 };
+	}
+
+	[[nodiscard]] ItemsContainerRange::Iterator end() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerRange::Iterator { nullptr, ContainerSize(), ContainerSize() };
+	}
+
+private:
+	[[nodiscard]] std::size_t ContainerSize() const
+	{
+		return sizeof(player_->SpdList) / sizeof(player_->SpdList[0]);
+	}
+
+	PlayerStruct *player_;
+};
+
+/**
+ * @brief A range over non-empty player items in the following order: Equipped, Inventory, Belt.
+ */
+class PlayerItemsRange {
+public:
+	explicit PlayerItemsRange(PlayerStruct &player)
+	    : player_(&player)
+	{
+	}
+
+	[[nodiscard]] ItemsContainerListRange::Iterator begin() const // NOLINT(readability-identifier-naming)
+	{
+		return ItemsContainerListRange::Iterator({
+		    EquippedPlayerItemsRange(*player_).begin(),
+		    InventoryPlayerItemsRange(*player_).begin(),
+		    BeltPlayerItemsRange(*player_).begin(),
+		});
+	}
+
+	[[nodiscard]] ItemsContainerListRange::Iterator end() const // NOLINT(readability-identifier-naming)
+	{
+		auto result = ItemsContainerListRange::Iterator({
+		    EquippedPlayerItemsRange(*player_).end(),
+		    InventoryPlayerItemsRange(*player_).end(),
+		    BeltPlayerItemsRange(*player_).end(),
+		});
+		return result;
+	}
+
+private:
+	PlayerStruct *player_;

```

Comment:
Do these variables have to be pointers? Looks like a reference should work since the constructor expects a reference.

Reply:
Pointers to make this class copyable